Add global HSFLL clock control device driver

bjarki-andreasen opened the pull request:
Draft for now since testing is blocked by missing firmware for the nRF54H20

ppelikan-nordic committed:
modules: hal_nordic: nrfs: add GDFS Service

Adding the implementation for the GDFS service

Signed-off-by: Paweł Pelikan <pawel.pelikan@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: nordic: specify device model of global hsfll clock

Add specific device model for global hsfll clock and update dts tree
to use specific model. The clock is not fixed, and configurable at
runtime to predefined frequencies specified by the platform.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: clock_control: nrf2: add support for global hfsll clock

Add device driver support for global hsfll clock.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
samples: boards: nordic: clock_control: support global hsfll

Extend sample to support global hsfll clock control.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: drivers: clock_control: nrf: add global hsfll clock

Extend test suite to test global HSFLL clock.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Note @ppelikan-nordic

bjarki-andreasen commented:
I need the to update the frequencies to match the opaque `GDFS_FREQ_MEDHIGH` in devicetree, and verify that the initial frequency of the clock is the lowest one :)

ppelikan-nordic commented on west.yml:
```diff
@@ -188,7 +188,7 @@ manifest:
         - hal
     - name: hal_nordic
-      revision: 54bde38c6f6ffb3780b26ae728cf79426184384e
+      revision: pull/254/head
```
Please change to https://github.com/zephyrproject-rtos/hal_nordic/pull/261 instead, to include newest fixes.
(https://github.com/zephyrproject-rtos/hal_nordic/pull/254 is closed now)

can be done by rebasing from: https://github.com/zephyrproject-rtos/zephyr/pull/81710

bjarki-andreasen-nordic committed:
manifest: hal_nordic: update to include nrfs services

Update hal_nordic revision to include nrfs services:
- GDPWR
- GDFS

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

gmarull committed:
soc: nordic: nrf54h: gpd: align GPD domain names

nRFs exposes now all power domains, following their actual name in the
specification. Add support for all of them in the GPD service. Note that
this is a breaking change: running this code requires a new SCFW as IDs
have changed in nRFs and so SCFW.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

57300 commented on dts/bindings/clock/nordic,nrf-global-hsfll.yaml:
```diff
@@ -0,0 +1,37 @@
+                                           320000000>;
+    };
+
+compatible: "nordic,nrf-global-hsfll"
```
I suggest `nordic,nrf-hsfll-global` to align with other Nordic bindings. Maybe even rename the local domain HSFLL binding to `nordic,nrf-hsfll-local`.

bjarki-andreasen replied:
I would like to do that, what about renaming `hsfll120:` to `hsfll_global:` in the devicetree?

57300 replied:
It can stay as `hsfll120` to match the datasheet.

bjarki-andreasen replied:
deal, and done

57300 replied:
Okay, you used `nordic,nrf-local-hsfll` instead?

bjarki-andreasen replied:
oh, sorry, did not notice the change of order :) will update

hubertmis replied:
note: there are multiple global HSFLLs: HSFLL120, 121, 200

bjarki-andreasen replied:
so far, only HSFLL120 is supported by GDFS, so any other clock is effectively a "fixed-clock" for now

hubertmis replied:
Sure, my comment was intended to guide on correct nodes, bindings, and compatibles naming

bjarki-andreasen replied:
Are there nodes which are misnamed currently that need to be addressed in this  PR?

hubertmis replied:
I don't know. I didn't review this PR. I only noticed you discussed renaming `hsfll120` to `hsfll_global` which would be correct only if HSFLL120 was the only global HSFLL instance. I shared a note that there are more instances to highlight that such renaming would be most probably incorrect.

bjarki-andreasen replied:
ah, makes sense :)

karstenkoenig replied:
> I would like to do that, what about renaming hsfll120: to hsfll_global: in the devicetree?

Sorry about being late, but I agree with @hubertmis here, calling hsfll120 the global hsfll is simplifying things too much. globalhsfll is used throughout the code here, in kconfig etc but only actually means hsfll120 and i think a mass rename from globalhsfll to hsfll120 would be way better

57300 commented on dts/bindings/clock/nordic,nrf-global-hsfll.yaml:
```diff
@@ -0,0 +1,37 @@
+
+  supported-clock-frequencies:
+    type: array
+    description: Supported clock frequencies in ascending order
```
Usually, the supported values are only described as an enum in the `clock-frequency` property binding. Hypothetically, if there was a DT macro for retrieving all possible enum values, would it make sense to use that in the driver, instead of inventing a new property?

bjarki-andreasen replied:
This property is distinct from `clock-frequency` given that the actual frequency is not known, it will be one of the supported frequencies, and it will vary. The user can't select any of them meaningfully in the devicetree, only describe which the driver can request, hence why I added this :)

bjarki-andreasen replied:
It has no `clock-frequency` binding since its not a fixed clock :)

57300 replied:
That makes sense. I guess `clock-frequency` could still signify the frequency requested at boot, but I see that Kconfig is used for that, so I assume that's more appropriate.

57300 commented on soc/nordic/nrf54h/global_hsfll.c:
```diff
@@ -0,0 +1,40 @@
+static struct onoff_client cli;
+static const struct device *global_hsfll = DEVICE_DT_GET(DT_NODELABEL(hsfll120));
+static const struct nrf_clock_spec spec = {
+	.frequency = CONFIG_SOC_NRF54H20_GDFS_MIN_FREQ,
```
nit: This file is created in the wrong commit, where it's in an unfinished state.

bjarki-andreasen-nordic committed:
dts: bindings: update nrf-hsfll to nrf-hsfll-local

The nrf-hsfll was previously the only supported HSFLL clock, hence it
was not namespaced fully. Since we added nrf-hsfll-global, we should
add the namespace to nrf-hsfll as well.

Updates drivers and devicetree uses of HSFLL as well.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: serial: nrfx_uarte: rm NRF_GPD_FAST_ACTIVE1 build assert

The assert BUILD_ASSERT(NRF_GPD_FAST_ACTIVE1 == 0); is not correct
given that NRF_GPD_FAST_ACTIVE1 is defined as 1U, and is not used
in the file anyway. Remove the build assert.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on west.yml:
```diff
@@ -189,7 +189,7 @@ manifest:
         - hal
     - name: hal_nordic
-      revision: pull/265/head
+      revision: fae15426c6b5a1f67362d508cf51b691ae5ab4b4
```
This should be done in the first commit, not here.

bjarki-andreasen replied:
given all three initial commits break if not applied together, the order is not really relevant here, I can squash them all together? its just not how I see it normally done :)

bjarki-andreasen replied:
wait, this looks wrong, let me check the commits

bjarki-andreasen replied:
fixed :)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -221,7 +208,6 @@ static DEVICE_API(nrf_clock_control, hsfll_drv_api) = {
 	.std_api = {
 		.on = api_nosys_on_off,
 		.off = api_nosys_on_off,
-		.get_rate = api_get_rate_hsfll,
```
Its not useful when the frequency is dynamic, it can change at any time, the only way to guarantee any frequency is to request what you need, and wait for confirmation, so there is no value in checking the freq like this at any time. Its also not possible with the global hsfll (not this clock I know) but it proves the point :) there we can't even get the freq if we wanted :)

anangl replied:
Agreed. It could be mentioned in the commit message so that the motivation is clear.

bjarki-andreasen commented on tests/drivers/clock_control/nrf_clock_control/src/main.c:
```diff
@@ -104,6 +104,37 @@ static const struct test_clk_context cpurad_hsfll_test_clk_contexts[] = {
+	{
+		.frequency = MHZ(320),
+		.accuracy = 0,
+		.precision = NRF_CLOCK_CONTROL_PRECISION_DEFAULT,
```
shortened

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -48,7 +47,7 @@ static const struct clock_options {
 
 struct hsfll_dev_data {
-	STRUCT_CLOCK_CONFIG(hsfll, ARRAY_SIZE(clock_options)) clk_cfg;
+	STRUCT_CLOCK_CONFIG(global_hsfll, ARRAY_SIZE(clock_options)) clk_cfg;
```
rogue mistake, changed back

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf-local-hsfll.yaml:
updated name of file to match compat

bjarki-andreasen commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -117,6 +120,11 @@ config NRFS_GDPWR_SERVICE_ENABLED
 	default y
 
+config NRFS_GDFS_SERVICE_ENABLED
+	bool "Global domain frequeny scaling service"
```
fixed

masz-nordic commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -37,6 +37,7 @@ zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RENESAS_RA_CGC      clock_cont
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_AMBIQ               clock_control_ambiq.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_PWM                 clock_control_pwm.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RPI_PICO            clock_control_rpi_pico.c)
+zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF2_GLOBAL_HSFLL   clock_control_nrf2_global_hsfll.c)
```
up

carlescufi reviewed (approved):
LGTM, thanks @bjarki-andreasen

masz-nordic commented on drivers/clock_control/clock_control_nrf2_global_hsfll.c:
```diff
@@ -0,0 +1,302 @@
+{
+	const struct global_hsfll_dev_config *dev_config = dev->config;
+
+	return dev_config->clock_frequencies[ARRAY_SIZE(dev_config->clock_frequencies) - 1];
```
This could be returned build-time since there are build asserts checking for supported clock frequencies?

bjarki-andreasen replied:
this driver does have some strangeness regarding single instances / multi instance code, I wanted to make it as easy as possible to extend it if we choose to support more HSFLLs in the future
